pkg/wasm/imports.go

```diff
@@ -0,0 +1,139 @@
+//go:build tinygo
+
+package main
+
+// This file contains JS functions imported into Go/WASM. Functions are
+// declared without body, the full definition can be found in the JS
+// implementation. The `jsRuntime` struct wraps these functions to
+// implement evaluator.Runtime.
+
+import (
+	"time"
+
+	"foxygo.at/evy/pkg/evaluator"
+)
+
+type jsRuntime struct {
+	yielder *sleepingYielder
+}
+
+func newJSRuntime() *jsRuntime {
+	return &jsRuntime{yielder: newSleepingYielder()}
+}
+
+func (*jsRuntime) Print(s string)                { jsPrint(s) }
+func (rt *jsRuntime) Read() string               { return rt.yielder.Read() }
+func (rt *jsRuntime) Sleep(dur time.Duration)    { rt.yielder.Sleep(dur) }
+func (rt *jsRuntime) Move(x, y float64)          { move(x, y) }
+func (rt *jsRuntime) Line(x, y float64)          { line(x, y) }
+func (rt *jsRuntime) Rect(x, y float64)          { rect(x, y) }
+func (rt *jsRuntime) Circle(r float64)           { circle(r) }
+func (rt *jsRuntime) Width(w float64)            { width(w) }
+func (rt *jsRuntime) Color(s string)             { color(s) }
+func (rt *jsRuntime) Yielder() evaluator.Yielder { return rt.yielder }
+
+type sleepingYielder struct {
+	start time.Time
+	count int
+}
+
+// newSleepingYielder yields the CPU so that JavaScript/browser events
+// get a chance to be processed. Currently(Feb 2023) it seems that you
+// can only yield to JS by sleeping for at least 1ms but having that
+// delay is not ideal. Other methods of yielding can be explored by
+// implementing a different yield function.
+func newSleepingYielder() *sleepingYielder {
+	return &sleepingYielder{start: time.Now()}
+}
+
+func (y *sleepingYielder) Yield() {
+	y.count++
+	if y.count > 1000 && time.Since(y.start) > 100*time.Millisecond {
+		time.Sleep(minSleepDur)
+		y.Reset()
+	}
+}
+
+func (y *sleepingYielder) Sleep(dur time.Duration) {
+	time.Sleep(dur)
+	y.Reset()
+}
+
+func (y *sleepingYielder) Read() string {
+	for {
+		if eval.Stopped {
+			return ""
+		}
+		addr := jsRead()
+		if addr != 0 {
+			return getStringFromAddr(addr)
+		}
+		y.Sleep(50 * time.Millisecond)
+	}
+}
+
+func (y *sleepingYielder) Reset() {
+	y.start = time.Now()
+	y.count = 0
+}
+
+// evySource is imported from JS. The float64 return value encodes the
+// ptr (high 32 bits) and length (low 32 bts) of the source string.
+//export evySource
```

Comment:
Same here and rest of file regarding blank comment lines and `goimports` on save.